### DISCUSSION TITLE
Joomla 4.2 & Multidomain

### DISCUSSION THREAD

**Discussion by olivdee** [joomla/joomla-cms#37846] [General] [answered]
Dear community,

we've been wondering why no info about the feature "multidomain in 4.2" has not been communicated lately (not found in "magazine" or "news" announcements about 4.2), and there seems not to be any PRs in here about it?
I suppose many users where eagerly expecting 4.2 because of this feature (at least we did), so an update somewhere if this goal has been discountinued would be great (maybe also with an info when/if this can be expected to re-enter the stage).

Thank you for a quick update and all the great work!

**Comment by HLeithner** [comment, marked as answer]
Sadly focus shifted and I and others hadn't the time to write this feature. Also customers who would sponsor this features didn't confirmed this yet.

So sorry to say no multi domain/multisite from my side for 4.2.

But we are still searching for help and would like to have it in core.

**Reply by olivdee** [reply]
Thanks for the reply. That is really unfortunate, as the way it was advertised/mentioned back in mid-21 many (as myself did) expected this to be happening... Well, it is the way it is, hopefully this feature gets added soon. All best & godspeed

**Reply by 4digitalworld** [reply]
yeah, this is the most anticipated feature of Joomla 4, I'm also looking forward to it

**Reply by simbus82** [reply]
It will be really a pity if even J5 will not follow the needs of the digital marketing & communication industry.
But it seems that with Joomla 5 ... fingers crossed!

**Comment by olivdee** [comment]
I just read that this will not be implemented even for J!5 in one year. Is this correct?

**Reply by brianteeman** [reply]
What does multidomain even mean?

**Reply by simbus82** [reply]
It is obvious that he is talking about multisite.
Today nearly all modern CMSes (and many really old) have a multidomain/multisite management. Even WP does it easily without extensions...

**Reply by brianteeman** [reply]
again what does it mean?
One install of joomla and 20 sites for clients that are all independent
One install of joomla and 20 identical sites on different domains
One install of Joomla and 20 sites sharing some of the content
Are all the sites independent or are they dependent on each other
Can you create content on each site or just on one

**Reply by simbus82** [reply]
The basis of a multisite is: a single installation of the CMS, many different sites each with its own domain, isolated from each other, with the selection of which configurative aspects to share.
A good old example is the inheritance that you can view on Magento, Prestashop, etc.
![image](https://user-images.githubusercontent.com/1375730/191570478-dae776c0-35cc-44e8-b079-458a3d7df3bd.png)

World is full of big companies with many brand and many markets... 
Joomla and other CMS are often not used precisely due to the impossibility of having a control from the HQ and then delegating the management of sub-brands to subsidiaries or to specific local management in the country.

This is why we often have to tie ourselves to systems such as SAP, Salesforce, Oracle, Adobe, etc. or worse to Shopify, Wix, etc ... which now provide multisite and headless CMS, often including PIM or DAM.

**Reply by brianteeman** [reply]
So if it is one installation and unlimited completely independent sites all isolated from each other what is the benefit. I can see benefis for ecommerce but not from content.
Not even one person asking for this has outlined how it could work and what are the benefits
It's not hard to do it but I dont see the benefits to justify the work.

However I strongly suspect that other peoples definition of multisite/domain is not the same as yours

**Reply by simbus82** [reply]
Buh, it's the same if you have to manage a network of blog/news/editorial website. Marketers that do linkbuilding (sell link and articles 😄) works a lot with this logic...

**Reply by Fedik** [reply]
Brian there many use of it: 
- corporate site with domain per region, per subdivision
- multilanguage domains: en.example.com, de.example.com, ua.example.com
- stuff like blog.example.com, shop.example.com, 
- category1.example.com, category2.example.com (for news, cataloge)
- etc

It not a big demand however.

**Reply by brianteeman** [reply]
Some of those are already possible. For others I still dont see why it would need to be one joomla install

**Reply by 4digitalworld** [reply]
> again what does it mean? One install of joomla and 20 sites for clients that are all independent One install of joomla and 20 identical sites on different domains One install of Joomla and 20 sites sharing some of the content Are all the sites independent or are they dependent on each other Can you create content on each site or just on one

What does this mean:

1) One installation of Joomla
2) Single login for all domains
3) Ability to choose what content will be published and on what domain (article, module, template, menu, etc.).

For example:
One article can be published on 2 out of 3 domains. (with a canonical link to one main domain)
One module can be published on 1 domain out of 3.
One template can be used by all 3 domains.
etc

**Reply by l-arnold** [reply]
Where this would be tremendously valuable is the following.

It takes a lot of work to fundamentally set up the control system for a single Joomla system.  If you are managing multiple sites you could do so from one console and spend less time assuring all systems are synchronized.  Also, as many extensions are only licensed for one system there would be cost savings with simply one system.

I have used the Virtual Domains system to do this with Joomla 3.x.  it is not perfect but those benefits are there if you remember always to add new content correctly also within the Virtual Domains module.b the fact is though it could be better than it is in various ways.

I also read that "Multiple Domains is coming" to Core Joomla.  It would be nice as it was the reason I was holding off my upgrade to Joomla 4...

I would be interested in helping to drive the spec.  I am not a PHP programmer and I am not 100% sure that this alone will cover the needs... (What about E-Commerce for instance)..

Anyway +1 for Multi Domains from my view.

**Comment by HDInfautre** [comment]
hello
Sad to learn that this is not in the roadmap joomla 5.
Well I saw that there are 2 extensions that would do this and then not sure if it will ever be compatible with a future multidomain joomla?

**Reply by brianteeman** [reply]
> Sad to learn that this is not in the roadmap joomla 5.

Where did you read that

**Reply by HDInfautre** [reply]
hi
> I just read that this will not be implemented even for J!5 in one year. Is this correct?
just above @olivdee
No statement to the contrary. So maybe a hope!?

**Reply by brianteeman** [reply]
https://volunteers.joomla.org/departments/production/reports/1859-production-dept-meeting-minutes-january-24-2023

**Reply by HDInfautre** [reply]
Thank you for this correct information and transparency.
I am also engaged in the francophone community as a non-developer

**Reply by gersteba** [reply]
On the page
https://volunteers.joomla.org/departments/production/reports/1859-production-dept-meeting-minutes-january-24-2023
it says "Multi domain / multi site implementation kick of is planned for the Cloudfest Hackthon in March".
What is the current state of that major feature?

**Comment by l-arnold** [comment]
Something in the works... that is good. 
I have been using "Virtual Domains" and it is not easy but it has endured. 

Look forward to what will be brought forward. 

Landis Arnold

**Comment by olivdee** [comment]
Since last week 2 big new projects started (2 hospital websites) with needed multidomain solutions (one language-multidomain and the other plain multidomain), and since I do not want to use AGAIN another CMS, I begun some experiments last weekend in order to achieve the needed functionality of multidomain handling without a 3rd party extension - because this feature is going to take a while until it arrives in core (as far as I understood). 

The article/solution of Marc (https://magazine.joomla.org/all-issues/december-2022/case-study-a-different-domain-name-for-each-language) is great/came very close to the needs of the first project, with one drawback, the language tags cannot be removed from the URL just by .htaccess rules and without coding a plugin "enhancing" the joomla router. 

Just by using the Joomla core language handling and the free plugin "n3t Language Domains", so far I could achieve the following:

1. **A different domain for each language**, with the language tag (/en/, /de/, ...) removed after the domain name.

2. **A different domain for different menus** (all have the same language).
This is a "misuse" of solution (1) with custom fictional installed languages, and slightly different setup in order to remove the alternate meta tags.

Both solutions can be implemented very quickly, no coding and no core hacks needed.
I guess this is a forwards compatible solution for the time when "multidomain" will be added in joomla core. 

Maybe this solution could also be interesting for the responsible dev-team? A different way of coming close to a core solution? I am just thinking out loud (maybe there is already a good planned solution): one could contact the n3t developers and talk to them if their plugin can be implemented in core - then "Languages" in the backend could be just renamed in "Languages & Multidomain" and both incorporated somehow together?
 
What I will try next, is if (2) "different domains for different menus" can also work with different languages.

**Questions:**

1. I did write a thorough tutorial for our company's internal wiki. Do you think it makes sense to post that in Joomla Community Magazine, so that others can benefit from it? If YES, what is the correct/best way to do this?

2. Does it make sense to post the 2 solutions somewhere else first (maybe here?) in order for one more into this to review it, in case I missed something out?

**Reply by crommie** [reply]
If you've decided what, when and how, feel free to write a JCM article. Here's how: https://magazine.joomla.org/all-issues/january-2023/how-to-write-for-the-joomla-community-magazine and https://magazine.joomla.org/all-issues/march-2023/submitting-your-articles-to-the-joomla-community-magazine

**Reply by olivdee** [reply]
Thanks! Will look into it soon. So, then somebody on that JCM team will do the quality management, reproduce/recheck my claims and give an approval?

**Reply by crommie** [reply]
If you want that to happen, write it in a Google doc first and have someone from Production look at it.

**Reply by olivdee** [reply]
@crommie hi and sorry for not replying, it has been and still is really busy since then ... please find our quick doc/instructions on my last answer - this is not close at all to a JCM article, but maybe one could use our notes to do so, or even help others to implement multidomain with (almost) only J!4 features

**Comment by FDE68** [comment]
Hello @olivdee - thank you for this info. 
Currently, we are searching also for the ability, to set up a multidomain page with joomla. In J3 there was an extension, which works fine - but it is not developed any more. 

So my question., are there any news to this?
As i know, it is not core of J4 or J5 ... how could i proceed, to get a little more knowledge, how to realize it? Could you help / send documentation? ... a few more infos?

Thank you!
Best regards, Frank

**Reply by olivdee** [reply]
Dear Frank, unfortunately we have been very busy since then, and hopefully can finish off two main projects by February. How urgent is this? I could try and give you the info needed to setup a basic multidomain system with Joomla Core mid/end of February. If you speak german, I could supply you with our internal notes to the test multidomain project even earlier. All best, Dino

**Reply by arnepluhar** [reply]
Hi Dino,
I'm also looking for this feature. I speak German and would love to receive your notes on this feature and also might be a candidate to test this functionality.
Best,
Arne

**Reply by FDE68** [reply]
Hello @olivdee - thank you for your answer. Yes, i am speaking german.
And currently we are in first planning of the page and some detais ... and we plan to start the setup in the middle of february. That could be a fit.
How to proceed to get more info?
What is needed? Should i start to setup a fresh page?
J4 or J5?
Thanks a lot for now.
Best Frank

**Reply by olivdee** [reply]
@FDE68 @arnepluhar please see my answer beneath, it was BTW already available in english

**Comment by olivdee** [comment]
Because I received two request for an insight into our solution, please find attached our tested and used (in Aug. 23) multilanguage solution with J!4. I am sorry that I might not be able to assist/answer questions in the next 2-3 weeks, but I can try if needed. Thanks for your understanding. Hopefully this will be helpful for some of you, or even give the J! team responsible for the multidomain implementation in core one day maybe a different approach?

[SOLUTIONS MULTILANGUAGE + MULTIDOMAIN.txt](https://github.com/joomla/joomla-cms/files/14112509/SOLUTIONS.MULTILANGUAGE.%2B.MULTIDOMAIN.txt)

**Comment by FDE68** [comment]
Hello @olivdee 
Thank you very much for this info ... so we can start to test ... 
Do i understand correctly, that if i want to realize the second part - different content / layout with different domains, i do not have to install the named plugin? ...plg_n3tlangdomains.4.0.2.zip ?
This is only for language per domain?

Or is this needed for both solutions?
If yes - where to get this plugin file?) (Link/page)

Thank you

**Comment by l-arnold** [comment]
Perhaps Helpful, 
Does the group have access to the code for "Virtual Domains". I should have it somewhere. 
Was working up through the last version of Joomla 3 for me. The code might be instructional. 

Mr. Mehrbrodt brought this forward from the original janguo-de build. 
This fork got the system working with PHP 7.1 and likely more details. 

https://github.com/smehrbrodt/virtualdomains 

I pulled a fork from his but have not changed anything. 
https://github.com/l-arnold/virtualdomains 

Anyway, there may be some details here worth looking at. 
I would certainly change somethings, but I did get this to work (from a functional, if not intuitive basis). 
The main problem I recall was the various ways you needed to keep things clean in your Joomla Backend. 
Needed to be a simple "drop down" for each domain so you would not get lost in your multiple Menu/Module configurations. 

I do have the server running still. Maybe 4 domains on it. They share some details, and don't share others. 

Ping me if you have questions, 
Landis Arnold
